### ISSUE TITLE
xgboost gpu predictor running slower relative to cpu

### ISSUE THREAD

**Issue by patelprateek** [dmlc/xgboost#3488]
I am training some gradient boosted trees using xgboost library. When i use the tree method : gpu_hist, i definitely see a 10x speedup for training. Another 4x speedup for using multiple gpus on a signle machine.

The issue I am running into right now is the prediction (using the gpu_predictor) is almost 1.5-2x slower than the cpu_predictor. Is this supposed to be like that ? I assumed the prediction/scoring can also be easily parallelized across all the boosted trees. Has any one faced similar issues ? Another issue i observe is that after training the on gpus , i still see memory not free'd on the gpus. Is this a known issue , do i need to delete the booster/model and run evaluation/prediction separately

**Comment by RAMitchell** [comment]
If you predict from a matrix that the algorithm has not seen before it needs to copy this matrix to the GPU - this can be slow. Our prediction algorithm is currently optimised for prediction during training as opposed to prediction from a 'warm start'.

And yes you will need to explicitly delete the booster to free the memory. Internally the algorithm has no way of knowing if the user has finished boosting or not so it must retain the memory.

**Comment by patelprateek** [comment]
@RAMitchell : what exactly you mean prediction during training ? is this something different than training itself ? I am using the same training dataset to run prediction to evaluate on training data.
What exactly is being compared here , rediction during training vs prediction warm start , can you elaborate further as this comparison doesnt make any sense to me . Prediction is just prediction , whether you run during training or a seprate eval dataset.
Also it doesn't seem memory copy to gpu is overhead here as i measure the time to copy to gpu is merely few seconds (100-1000x smaller than actual prediction time).

**Comment by patelprateek** [comment]
Even after deleting the booster and reading back serialized model and doing some warmpup to run prediction , the prediction is much slower than cpu , this after removing the memory copy overhead.
What exactly is optimized in gpu predictor as opposed to cpu predictor ? I am trying different dataset size and the result is pretty consistent with gpu predictor always being worse than cpu predictor. training takes about 2-3 minutes whereas prediction is taking 30 minutes , i don't think this is something expected.  any workarounds

**Comment by hcho3** [comment]
> training takes about 2-3 minutes whereas prediction is taking 30 minutes

Do you have a re-producible example for this?

**Comment by patelprateek** [comment]
I do not have a public dataset , but i can give you 50-80 features , data size 2Gb , 2-3 million rows . I also tried with much larger dataset 33 million rows around 24 Gb

**Comment by hcho3** [comment]
Let me ask you a few more questions:
* Which version of XGBoost are you using? If compiled from the source, what is the git commit hash?
* How many trees does the model have? What is the average number of leaves and depth for each tree?
* Which model of NVIDIA GPU are you using? How much memory does it have?

> I do not have a public dataset

Without the data it would be difficult for us to diagnose the cause for low performance. Can you generate a dummy data that has similar characteristics as your dataset (in number of features, number of nonzero entries, number of data rows, scale of each feature etc.)?
